core/src/cost_update_service.rs

```diff
@@ -39,7 +46,27 @@ impl CostUpdateService {
         for cost_update in cost_update_receiver.iter() {
             match cost_update {
                 CostUpdate::FrozenBank { bank } => {
-                    bank.read_cost_tracker().unwrap().report_stats(bank.slot());
+                    for loop_count in 0..MAX_LOOP_COUNT {
+                        // Release the lock so that the thread that will update
+                        // the count is able to obtain a write lock
+                        let in_flight_transaction_count = bank
+                            .read_cost_tracker()
+                            .unwrap()
+                            .in_flight_transaction_count();
+                        if in_flight_transaction_count == 0 {
```

Comment:
Do we know how often in flight transaction count is above zero when we come in here? Like is this happening most of the time?

If in flight is 0 most of the time, I'm wondering if we can restructure things to only acquire the read lock once for that typical case. Something like:
```
                    loop {
                        {
                            let cost_tracker = bank.read_cost_tracker().unwrap();
                            let in_flight_transaction_count = cost_tracker
                                .in_flight_transaction_count();
                            if in_flight_transaction_count == 0 || loop_count >= MAX_LOOP_COUNT {
                                let slot = bank.slot();
                                trace!(
                                    "inflight transaction count for slot {slot} at {in_flight_transaction_count} \
                                    after {loop_count} iterations"
                                );
                                cost_tracker.report_stats(bank.slot());
                            }
                        }

                        std::thread::sleep(LOOP_LIMITER);
                    }
```

Reply:
I don't think we have numbers for how often, but in any case, I agree with your sentiment that we can rework to avoid repeated lock grabs. I'll try to work this in.

Unrelated but as an FYI - GH allows you do syntax highlighting for code blocks. To do so, type out the following
> \```rust
> desired code block
> ...
> \```

For example,

```rust
loop {
    // Grab the lock once here
    let cost_tracker = bank.read_cost_tracker().unwrap();
```